emulator/src/kreator/preprocessor.rs

```diff
@@ -180,18 +190,149 @@ fn replace_macros(code: &Vec<String>) -> Result<Vec<String>, &'static str> {
+        let mut owned_line = line.clone();
+
+        // check if current block of code is (not) a macro
+        if owned_line.eq(MACRO_START) {
+            in_macro = true;
+            continue;
+        } 
+        if owned_line.eq(MACRO_END) {
+            in_macro = false;
+            label_map.clear();
+            equ_map.clear();
+            set_map.clear();
+            continue;
+        }
+
+        // find set assignments outside of macros
+        if owned_line.contains(" SET ") && !in_macro {
+            let (name, _) = owned_line.split_once(" SET ").unwrap();
+            found_set_names.push(name.to_string());
+            all_existing_names.push(name.to_string());
+        }
+
+        // change line of macro according to principles of locality
+        if in_macro {
+            
+            // map local labels in macros
+            if loc_label_regex.is_match(&owned_line) && !glob_label_regex.is_match(&owned_line) {
+                let (label, _) = owned_line.split_once(":").unwrap();
+                let gen_name = generate_label_name(&all_existing_names);
+                all_existing_names.push(gen_name.clone());
+                label_map.insert(label.to_string(), gen_name);
+            }
+
+            // convert globally declared label to normal label
+            if glob_label_regex.is_match(&owned_line) {
+                owned_line = owned_line.replace("::", ":");
+            }
+
+            // map local equ assignments
+            if line.contains(" EQU ") {
+                let (name, _) = owned_line.split_once(" EQU ").unwrap();
+                let gen_name = generate_label_name(&all_existing_names);
+                all_existing_names.push(gen_name.clone());
+                equ_map.insert(name.to_string(), gen_name);
+            }
+
+            // replace local label calls and equ assigned variables
+            for (old_label, generated_label) in &label_map {
+                owned_line = owned_line.replace(old_label, generated_label);
+            }
+            for (old_name, new_name) in &equ_map {
+                owned_line = owned_line.replace(old_name, new_name);
+            }
+            
+            // check if set assignment variable existed outside of macro
+            if owned_line.contains(" SET ") {
+                let (name, _) = owned_line.split_once(" SET ").unwrap();
+                if !found_set_names.contains(&name.to_string()) {
+                    let gen_name = generate_label_name(&all_existing_names);
+                    all_existing_names.push(gen_name.clone());
+                    set_map.insert(name.to_string(), gen_name);
+                }
+            }
+            for (old_name, new_name) in &set_map {
+                owned_line = owned_line.replace(old_name, new_name);
+            }
+        }
+        handled_code.push(owned_line.to_string());
+    }
+    Ok(handled_code)
+}
+
+fn generate_label_name(taken_names: &Vec<String>) -> String {
```

Comment:
I will need a very good explanation, why label names are generated randomly

Reply:
I'm not sure if it matters in this context, but I chose this implementation to have the maximum of labels available, which, in hindsight, might not be needed. Of course we could simply name them something like 'GEN00' counting upwards, depending on the name the amount is restricted more heavily in this case.

The only goal in this case is to get a name that has not been assigned yet, I don't think it is too important what the name is as it will only be seen by the preprocessor, not the user.

Reply:
As for the restriction part: Keep in mind that whatever the calling method produces will exclusively be handled by the preprocessor class, meaining the produced code has to be legal. This means all labels have to be legal labels (start with a letter or @, be 1-5 characters in length), making it impossible to simply count up the label names. Additionally we cannot count above (CHAR)9999.

I still don't want to simply dismiss your objection as it is a valid claim and I think testing would benefit from a better naming scheme. However I gravely underestimated the amount of options available (and made another error when thinking of the implementation of all of this).

I would propose to name labels A0000 to Z9999, which leaves us with 260000 possible unique labels whenever a macro is called. If this does not suffice we could  switch to counting in hex to further raise the amount (although I don't think anyone would ever need that many)

Reply:
Yeah that sounds great